treegp/two_pcf.py

```diff
@@ -0,0 +1,410 @@
+    :param max_sep:     Maximum bin for treecorr. (float)
+    :param nbins:       Number of bins (if 1D correlation function) of the square root of the number
+                        of bins (if 2D correlation function) used in TreeCorr to compute the
+                        2-point correlation function. [default: 20]
+    :param anisotropic: 2D 2-point correlation function.
+                        Used 2D correlation function for the
+                        fiting part of the GP. (Boolean)
+    :param robust_fit:  Used minuit to fit hyperparameter. Works only
+                        anisotropic is True. (Boolean)
+    """
+    def __init__(self, X, y, y_err,
+                 min_sep, max_sep, nbins=20, 
+                 anisotropic=False, robust_fit=False, 
+                 p0=[3000., 0., 0.]):
+        self.ndim = np.shape(X)[1]
+        if self.ndim not in [1, 2]:
+            raise ValueError('two-pcf support only 1d and 2d modeling for the moment. curent ndim: %i'%(self.ndim))
+        if self.ndim == 2:
+            self.X = X
+        if self.ndim == 1:
+            self.X = np.array([X.T, np.zeros_like(X.T)]).T[:,0]
+        self.y = y
+        self.y_err = y_err
+        self.min_sep = min_sep
+        self.max_sep = max_sep
+        self.nbins = nbins
+        self.anisotropic = anisotropic
+        self.robust_fit = robust_fit
+        self.p0_robust_fit = p0
+
+    def resample_bootstrap(self):
+        """
+        Make a single bootstrap resampling on data.
+        """
+        npsfs = len(self.y)
+
+        ind_object = np.random.randint(0,npsfs-1, size=npsfs)
+        u_ressample = self.X[:,0][ind_object]
+        v_ressample = self.X[:,1][ind_object]
+        y_ressample = self.y[ind_object]
+        y_err_ressample = self.y_err[ind_object]
+
+        return u_ressample, v_ressample, y_ressample, y_err_ressample
+
+    def comp_2pcf(self, X, y, y_err):
+        """
+        Estimate 2-point correlation function using TreeCorr.
+
+        :param X:  Coordinates of the field. (n_samples, 1 or 2)
+        :param y:  Values of the field. (n_samples)
+        :param y_err: Error of y. (n_samples)
+        """
+        if np.sum(y_err) == 0:
+            w = None
+        else:
+            w = 1./y_err**2
+
+        if self.anisotropic:
+            cat = treecorr.Catalog(x=X[:,0], y=X[:,1], k=(y-np.mean(y)), w=w)
+            kk = treecorr.KKCorrelation(min_sep=self.min_sep, max_sep=self.max_sep, nbins=self.nbins,
+                                        bin_type='TwoD', bin_slop=0)
+            kk.process(cat)
+            # Need a mask in the case of the 2D correlation function, to compute
+            # the covariance matrix using the bootstrap. The 2D correlation
+            # function is symmetric, so just half of the correlation function
+            # is useful to compute the covariance matrix. If it is not done,
+            # the covariance matrix is consequently not positive definite.
```

Comment:
I wonder how the speed of computing the 2pcf via treecorr compares to directly using FFTs... I might have to test this at some point.

Reply:
This is a good point and you might be right. I remember that Pierre Astier in [this article](https://www.aanda.org/articles/aa/full_html/2019/09/aa35508-19/aa35508-19.html) is computing an anisotropic 2-point correlation function of the brighter-fatter effect (Appendix A). I remember he told me it was fast, but how it compare to `treecorr`, I don't know.